### PR TITLE
encryption: pipeline save file dict

### PR DESCRIPTION
**Pull request by Little-Wallace**
Signed-off-by: Little-Wallace <bupt2013211450@gmail.com>

### What problem does this PR solve?

Problem Summary:
Encryption manager has to save the file dict on disk every time when there is a new file created or updated. It will cost too much time to do this operation. https://github.com/tikv/tikv/pull/8865 has solved thie problem but it is hard to pick this change to release-4.0 because it also changes the file format, if user upgrade one cluster from release-4.0 to https://github.com/tikv/tikv/pull/8865, he can not go back to release-4.0 again. 

### What is changed and how it works?

This PR will make all of thread which want to change file dict in a pipeline queue. In origin implement, if there is  4 thread trying to call `save_file_dict`,  it may wait 3 * IO round and cost one IO round to sync the file dict again. But it is not necessary, because the second thread can sync the result of the 3rd and the 4th thread together, the 3rd and the 4th only need to wait 2 * IO round and cost no time to sync itself changes.
In this PR, no matter how many threads there are trying to call `save_file_dict`,  every of them only need to wait 2 IO round and do nothing at most.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- No release note.

### CODE REVIEW

File: components/encryption/src/manager/mod.rs [hunk 1]
```diff
@@ -18,16 +20,21 @@ use crate::io::EncrypterWriter;
...
 
 const KEY_DICT_NAME: &str = "key.dict";
 const FILE_DICT_NAME: &str = "file.dict";
 const ROTATE_CHECK_PERIOD: u64 = 600; // 10min
+const DEFAULT_FILES_CACHE_CAPACITY: usize = 1024 * 128;
 
 struct Dicts {
     // Maps data file paths to key id and metadata. This file is stored as plaintext.
     file_dict: Mutex<FileDictionary>,
+    files_cache: DashMap<String, FileInfo>,
```
Comment (hunterlxt) [inline components/encryption/src/manager/mod.rs:34]: Add comments for this?

File: components/encryption/src/manager/mod.rs [hunk 2]
```diff
@@ -175,31 +190,80 @@ impl Dicts {
...
+            }
+            assert!(write_queue.len() > 0);
+            write_queue.len()
+        };
+        let ret = self.save_file_dict();
+        let save_err = ret.is_err();
+        if let Err(e) = ret {
+            error!("encryption: failed to save file dictionary."; "error" => ?e);
+        }
+        let mut write_queue = self.write_queue.lock().unwrap();
```
Comment (hunterlxt) [inline components/encryption/src/manager/mod.rs:243]: The logic is complicated... I read the part several times, it should be okay. @yiwu-arbug Pls check again.

Comment (Little-Wallace) [reply components/encryption/src/manager/mod.rs:243]: I will add some comment to make it easy to understand.

**Comment by yiwu-arbug** [comment]
I incline to just cherry-pick #8865 in a somehow -forward-compatible way. The pipelined approach is error-prone and don't bring much value - we only need it for 4.0 release.

**Comment by Little-Wallace** [comment]
We will cherry-pick  #8865 to solve this problem because the logic of this PR is too complicated.

**Pull request closed without merging**